### PR TITLE
FilenameBear.py: Add filename length check

### PR DESCRIPTION
**Pull request by damian1996**
Adds the check if filename has length more than 4096 on UNIX
systems or more than 260 on Windows.
In that case, user will get message about it. 

Closes https://github.com/coala/coala-bears/issues/2003

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by Makman2** [comment]
Commit message:
I would slightly rephrase the shortlog: `Add filename length check`
And the body is wrong: If the length has less than 4096, it's actually okay. And nothing gets shortened there from coala side, the user has to do it.

File: bears/general/FilenameBear.py [hunk 1]
```diff
@@ -98,6 +98,11 @@ def run(self, filename, file,
                 'Filename does not use the suffix {!r}.'.format(
                     filename_suffix))
 
+        if len(new_name) > 4096-len(extension):
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: Have you thought about making this a setting instead of hardcoding the value? Also what about the 260 char limit of windows? Shall the default limit be maybe system dependent?

**Comment by Makman2** [comment]
unack 58ddd11

File: bears/general/FilenameBear.py (hunk 1 above)
Comment (damian1996) [reply bears/general/FilenameBear.py]: Additionally, you metioned about system dependent limit and I tried to check limits on Windows, Linux and OS X. I found that limit on Windows is really 260, limit on Linux is 4096 and also limit on OS X is different than previous metioned (probably 255). So do you think that better idea would be to check the current system or set default length as minimum of all systems?

Comment (Makman2) [reply bears/general/FilenameBear.py]: Probably, at least the default should may be variable. However for the sake of cross-compatibility this might be less good.

> OS X is different than previous metioned (probably 255)

Should be something Linux-like, as OSX is more or less BSD.

> I found that limit on Windows is really 260

And I believe it's possible to change the limit in Windows. 260 chars is for compatibility with older software (and sadly still default). Maybe we should detect the actual compatibility mode? Not sure, sounds like work 😅

Comment (damian1996) [reply bears/general/FilenameBear.py]: Should I consider that on Linux (and probably OSX) maximum length of filename is 255, however maximum path length is even 4096?

Comment (damian1996) [reply bears/general/FilenameBear.py]: Additionally, [here](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx#maxpath) is metioned that maximum path length is 260, however.. 
"The Windows API has many functions that also have Unicode versions to permit an extended-length path for a maximum total path length of 32,767 characters." 
Seems be that this is various in dependent of functions, however this can be hard to detection, so better choice should be take minimum, which is 260.

Comment (Makman2) [reply bears/general/FilenameBear.py]: > so better choice should be to take minimum, which is 260.

sounds good :)

File: bears/general/FilenameBear.py [hunk 2]
```diff
@@ -26,6 +27,8 @@ class FilenameBear(LocalBear):
                                   }
 
     def run(self, filename, file,
+            max_len_path_unix: int=4096,
+            max_len_path_windows: int=260,
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: Ah interesting, separating this for different systems. I'm just considering if it still can be made a single settings, I'm thinking of
```
max_path_length = 4096  # <-- affect all systems
max_path_length = {'windows': 100, 'unix': 100} # <-- different settings for different OSes, but this is probably too complicated from user perspective
```
However I think this can be a new issue.

Comment (Makman2) [inline bears/general/FilenameBear.py]: Just thinking, do you mean path length or filename length? Paths can be different on different environments.

Also:
Change name pattern to `max_path_length_...` instead of `max_len_path_...`

File: bears/general/FilenameBear.py [hunk 3]
```diff
@@ -98,6 +101,13 @@ def run(self, filename, file,
                 'Filename does not use the suffix {!r}.'.format(
                     filename_suffix))
 
+        p = platform.system()
+        max_len = max_len_path_windows if p == 'Windows' else max_len_path_unix
+        if len(new_name) > max_len-len(extension):
+            messages.append(
+                'Filename length is more than {}, so it is too long.'.format(
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: 1. you should be precise with path vs filename
2. incorporate the actual length, I would recommend
    ```
    Filename length ({}) is more than {}.
    ```
    or even better imo
    ```
    'Filename is too long ({} > {})'
    ```
    but up to you to phrase a nice message ;)
3. "so it is too long" is redundant ;)

Comment (Makman2) [reply bears/general/FilenameBear.py]: Oh maybe you should also incorporate the system the user is on, so he isn't confused when he sees "Wait, my filename is too long? But `filename_length_unix` is 1000", and because he's on Windows he oversees that the windows-default is used ;) (unlikely, but just a suggestion to provide more information in the message)

File: tests/general/FilenameBearTest.py [hunk 4]
```diff
@@ -151,3 +152,25 @@ def test_file_prefix_suffix(self):
...
+        if platform.system() == 'Windows':
+            filename_testWin1 = '_filenamewhichistoolong'*15+'.xyz'
+            filename_testWin2 = '_filenamewhichistoolong'*10+'.xyz'
+            filename_testWin3 = '_validfilenamelength'*13
+            self.check_invalidity(
+                self.uut, [''], filename=filename_testWin1)
+            self.check_validity(
+                self.uut, [''], filename=filename_testWin2)
+            self.check_validity(
+                self.uut, [''], filename=filename_testWin3)
```
Comment (Makman2) [inline tests/general/FilenameBearTest.py]: could you use `check_results`? I want to see tests for the message :+1:

**Comment by Makman2** [comment]
unack e9db5e7

File: bears/general/FilenameBear.py (hunk 2 above)
Comment (damian1996) [reply bears/general/FilenameBear.py]: Should I check only filename length or filename and path length?

**Comment by damian1996** [comment]
I think, that limit filename length should be 255 bytes on Linux, because 4096 is other limit (path length).

File: bears/general/FilenameBear.py (hunk 3 above)
Comment (damian1996) [reply bears/general/FilenameBear.py]: I think, that this message is great :) 
> 'Filename is too long ({} > {})'

File: bears/general/FilenameBear.py (hunk 2 above)
Comment (damian1996) [reply bears/general/FilenameBear.py]: What do you think about such situation:
When we have different limits on different systems, then when project is moved to other system, then this code wouldn't be useful. Because on first system limit was 260, but on second system is only 255. 
In my opinion setting the minimum of these limits as default should resolve it.

**Comment by damian1996** [comment]
What do you think about such situation:
When we have different limits on different systems, then when project is moved to other system, then this code wouldn't be useful. Because on first system limit was 260, but on second system is only 255.
In my opinion setting the minimum of these limits as default should resolve it.

**Comment by Makman2** [comment]
> In my opinion setting the minimum of these limits as default should resolve it.

agree :+1: :)

File: bears/general/FilenameBear.py [hunk 5]
```diff
@@ -42,6 +43,9 @@ def run(self, filename, file,
             - ``pascal`` (``ThisIsPascalCase``)
             - ``snake`` (``this_is_snake_case``)
             - ``space`` (``This Is Space Case``)
+        :param max_filename_length:
+            Minimum of maximum filename length on Windows and maximum filename
+            length pn Unix-like systems. Default value is 260 characters.
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: typo: pn -> on
\+
No need to mention the default value, coala automatically does that.

File: bears/general/FilenameBear.py [hunk 6]
```diff
@@ -42,6 +43,9 @@ def run(self, filename, file,
             - ``pascal`` (``ThisIsPascalCase``)
             - ``snake`` (``this_is_snake_case``)
             - ``space`` (``This Is Space Case``)
+        :param max_filename_length:
+            Minimum of maximum filename length on Windows and maximum filename
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: > Minimum of maximum

that sounds confusing :3 Either minimum or maximum (I tend to 2nd ;D)

File: bears/general/FilenameBear.py [hunk 7]
```diff
@@ -54,7 +58,6 @@ def run(self, filename, file,
         """
         head, tail = os.path.split(filename)
         filename_without_extension, extension = os.path.splitext(tail)
-
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: unnecessary change

File: bears/general/FilenameBear.py [hunk 8]
```diff
@@ -98,14 +101,19 @@ def run(self, filename, file,
...
+                    max_filename_length))
+
         if ignore_uppercase_filenames and filename_without_extension.isupper():
             return
 
         if messages:
             diff = Diff(file, rename=os.path.join(head, new_name + extension))
             message = ('\n'.join('- ' + mes for mes in messages)
                        if len(messages) > 1 else messages[0])
-
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: unnecessary change

File: bears/general/FilenameBear.py [hunk 9]
```diff
@@ -98,14 +101,19 @@ def run(self, filename, file,
                 'Filename does not use the suffix {!r}.'.format(
                     filename_suffix))
 
+        if len(new_name+extension) > max_filename_length:
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: I prefer spaces around operators, but optional :+1:

Comment (Makman2) [inline bears/general/FilenameBear.py]: Hm okay you are analyzing the new name generated. We should be careful not to confuse the user here, because he thinks "huh, but my file isn't called like that / the length of the name doesn't match up :O".

File: bears/general/FilenameBear.py [hunk 10]
```diff
@@ -98,14 +101,19 @@ def run(self, filename, file,
...
+            messages.append(
+                'Filename is too long ({} > {}).'.format(
+                    len(new_name+extension),
```
Comment (Makman2) [inline bears/general/FilenameBear.py]: also here but optional: spaces around operator

File: bears/general/FilenameBear.py (hunk 9 above)
Comment (damian1996) [reply bears/general/FilenameBear.py]: What do you think about adding whole name of file (I mean new_name + extension) in error message? :)

Comment (Makman2) [reply bears/general/FilenameBear.py]: still not sure...

Imo I would use the "old" filename for this check, I think even displaying `new_name + extension` would confuse the user. For now I would say we just use the old name, if users start to complain about this setting, we think about it again ;)

Comment (damian1996) [reply bears/general/FilenameBear.py]: I'm still not sure. You should imagine such scenario : 
If we'd check filename on the start, then we aren't catch too long filename after adding required suffix or prefix (this is possible).
So in my opinion checking this on new_name is better solution.

**Comment by Makman2** [comment]
ack bd08e1b

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
@gitmate-bot ff
